django/db/models/options.py

```diff
@@ -982,6 +982,30 @@ def _property_names(self):
                 names.append(name)
         return frozenset(names)
 
+    @cached_property
+    def _concrete_field_names(self):
+        """Return a set of the concrete field names defined on the model."""
+        names = []
+        for field in self.concrete_fields:
+            if not field.primary_key:
+                names.append(field.name)
+
+                if field.name != field.attname:
+                    names.append(field.attname)
+        return frozenset(names)
+
+    @cached_property
+    def _auto_update_field_names(self):
+        """Return a set of field names which auto update on model save."""
+        names = []
+        for field in self.concrete_fields:
+            if getattr(field, "auto_now", False):
```

Comment:
Is `auto_now` the only case where we have this behaviour?
Would there be a way to detect if a user defined a custom model field that would also have a similar behaviour, but doesn't have an `auto_now`?
Looking at some implementations in Django, `django.db.models.fields.files.FileField` is by default saving the file if not committed yet. We perhaps need to handle those cases too? 🤔

Reply:
You're completely right, I have made some edits to move it away from specifically checking for auto_now fields and to instead add any field which is edited on the fields pre_save (which includes the auto_now fields)
Let me know what you think :+1: